middleware/glcd/lib/src/glcd.c

```diff
@@ -0,0 +1,786 @@
+
+                        uint8_t page = cy / 8;
+                        uint8_t bit_in_page = 1 << (cy % 8);
+                        uint8_t val = glcd_read(glcd, page, x);
+                        glcd_write(glcd, page, x, val | bit_in_page);
+                    }
+                }
+                break;
+            }
+
+            case DIAGONAL:
+            {
+                int x0 = p0.x, y0 = p0.y;
+                int x1 = p1.x, y1 = p1.y;
+                int dx = abs(x1 - x0), dy = abs(y1 - y0);
+                int sx = (x0 < x1) ? 1 : -1;
+                int sy = (y0 < y1) ? 1 : -1;
+                int err = dx - dy;
+
+                while (1)
+                {
+                    for (uint8_t dy_dot = 0; dy_dot < thickness; dy_dot++)
+                    {
+                        for (uint8_t dx_dot = 0; dx_dot < thickness; dx_dot++)
+                        {
+                            int x = x0 + dx_dot;
+                            int y = y0 + dy_dot;
+                            if (x < 0 || x >= 128 || y < 0 || y >= 64) continue;
+
+                            uint8_t page = y / 8;
+                            uint8_t bit_in_page = 1 << (y % 8);
+                            uint8_t val = glcd_read(glcd, page, x);
+                            glcd_write(glcd, page, x, val | bit_in_page);
+                        }
+                    }
+
+                    if (x0 == x1 && y0 == y1) break;
+
+                    int e2 = 2 * err;
+                    if (e2 > -dy) { err -= dy; x0 += sx; }
+                    if (e2 < dx)  { err += dx; y0 += sy; }
+                }
+                break;
+            }
+        }
+    }
+}
+
+void glcd_draw_rect(glcd_t* glcd, const point* p, uint16_t psize, const rect* r, uint16_t rsize)
+{
+    if (!glcd || !r || !p || psize != rsize ) return;
+
+    for (uint16_t i=0; i<psize; i++)
+    {
+        uint8_t x0 = p[i].x, y0 = p[i].y;
+        uint8_t x1 = (x0 + r[i].w)%ROW_SIZE, y1 = (y0 + r[i].h)%COL_PER_CHIP;
+        uint8_t line_sz = r[i].line_size;
+
+        segment rect[4] = {
+            {{{x0, y0},{x1, y0}}, line_sz},
+            {{{x0, y0},{x0, y1}}, line_sz},
+            {{{x1, y0},{x1, y1}}, line_sz},
+            {{{x0, y1},{x1, y1}}, line_sz}
+        };
+
+        glcd_draw_line(glcd, rect, 4, DIAGONAL);
+        if (r[i].filled) { fill_polygon(glcd, rect, 4); }
+        if (r[i].rounded) { /*TODO*/ }
+    }
+}
+
+void glcd_draw_shape(glcd_t* glcd, const segment* edges, uint8_t size, bool is_filled, bool round_edges)
+{
+    if (!glcd || !edges) return;
+
+    segment output[64];
+    sort_points_nearest_neighbor(edges, output, size);
+    glcd_draw_line(glcd, output, size, DIAGONAL);
+
+    if (is_filled) { fill_polygon(glcd, output, output[0].line_size); }
+    if (round_edges) { /*TODO*/ }
+}
+
+void glcd_draw_regular_polygon(glcd_t* glcd, const point* ori, uint8_t num_of_ori, const polygon_mode_t* pol, uint8_t num_of_pol, bool is_filled)
```

Comment:
```suggestion
void glcd_draw_regular_polygon ( glcd_t* glcd, const point* ori,
                                 uint8_t num_of_ori, const polygon_mode_t* pol,
                                 uint8_t num_of_pol, bool is_filled )
```
Use spaces inside of brackets throughout the whole code as in this code suggestion.
Make sure the lines don't exceed the length of 120 charactes (ideally <100).